File: bandlog-api/src/main/java/net/effize/bandlog/domain/user/model/Email.java
package net.effize.bandlog.domain.user.model;

import jakarta.persistence.Embeddable;
import net.effize.bandlog.domain.user.exception.EmailMalformedException;

@Embeddable
public record Email(
        String value
) {
    public Email {
        if (!isValid(value)) {
            throw new EmailMalformedException();
        }
    }

    public String fullEmail() {
        return value;
    }

    private boolean isValid(String email) {
        return email != null &&
                email.matches("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$");
    }
}

File: src/main/kotlin/net/effize/bandlog/user/domain/Email.kt
package net.effize.bandlog.user.domain

import jakarta.persistence.Embeddable
import net.effize.bandlog.user.domain.exception.EmailMalformedException

@Embeddable
data class Email(
    val value: String
) {
    init {
        require(isValid(value)) { throw EmailMalformedException() }
    }

    fun fullEmail(): String = value

    private fun isValid(email: String?): Boolean {
        return email != null &&
                email.matches(Regex("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"))
    }
}
